packages/mongo/collection.js

```diff
@@ -455,51 +455,99 @@ Object.assign(Mongo.Collection.prototype, {
+            throw new Error(
+              "Meteor requires document _id fields to be non-empty strings or ObjectIDs");
+          }
+        } else {
+          // We need to generate the _id for the document
+          let generateId = true;
+
+          // Don't generate the id if we're the client and the 'outermost' call
+          // This optimization saves us passing both the randomSeed and the id
+          // Passing both is redundant.
+          if (this._isRemoteCollection()) {
+            const enclosing = DDP._CurrentMethodInvocation.get();
+            if (!enclosing) {
+              generateId = false;
+            }
+          }
 
-    if ('_id' in doc) {
-      if (! doc._id ||
-          ! (typeof doc._id === 'string' ||
-             doc._id instanceof Mongo.ObjectID)) {
-        throw new Error(
-          "Meteor requires document _id fields to be non-empty strings or ObjectIDs");
-      }
+          if (generateId) {
+            doc._id = this._makeNewID();
+          }
+        }
+      });
+      doc = insertDocs;
     } else {
-      let generateId = true;
-
-      // Don't generate the id if we're the client and the 'outermost' call
-      // This optimization saves us passing both the randomSeed and the id
-      // Passing both is redundant.
-      if (this._isRemoteCollection()) {
-        const enclosing = DDP._CurrentMethodInvocation.get();
-        if (!enclosing) {
-          generateId = false;
+      // Make a shallow clone of the document, preserving its prototype.
+      doc = Object.create(
+        Object.getPrototypeOf(doc),
+        Object.getOwnPropertyDescriptors(doc)
+      );
+
+      if ('_id' in doc) {
+        if (! doc._id ||
+          ! (typeof doc._id === 'string' ||
+            doc._id instanceof Mongo.ObjectID)) {
+          throw new Error(
+            "Meteor requires document _id fields to be non-empty strings or ObjectIDs");
+        }
+      } else {
+        let generateId = true;
+
+        // Don't generate the id if we're the client and the 'outermost' call
+        // This optimization saves us passing both the randomSeed and the id
+        // Passing both is redundant.
+        if (this._isRemoteCollection()) {
+          const enclosing = DDP._CurrentMethodInvocation.get();
+          if (!enclosing) {
+            generateId = false;
+          }
         }
-      }
 
-      if (generateId) {
-        doc._id = this._makeNewID();
+        if (generateId) {
+          doc._id = this._makeNewID();
+        }
       }
     }
 
+
+
     // On inserts, always return the id that we generated; on all other
     // operations, just return the result from the collection.
-    var chooseReturnValueFromCollectionResult = function (result) {
+    const chooseReturnValueFromCollectionResult = function (result) {
+      if (isBulkInsert) {
+        console.dir(result)
+        if (result.insertedIds) {
+          return Object.values(result.insertedIds);
```

Comment:
What is the purpose of `Object.values` here? (Also a stray `console` call, if you weren't already aware.)

Reply:
Yes, stray condole. The purpose is get the ids as the return format from MongoDB is like: `{ "0": "idOfTheDocument", "1": "idOfTheDocument" }`